### ISSUE TITLE
ISelectiveResourceReloadListener

### ISSUE THREAD

**Issue by maxanier** [MinecraftForge/MinecraftForge#6553]
ISelectiveResourceReloadListener is not server safe https://github.com/MinecraftForge/MinecraftForge/issues/5837

This itself is not an issue, but if some mod uses it anyway it prevents MC from loading data (packs) and it is not possible to trace this back to that mod.
```
[06Mar2020 06:36:06.861] [Server thread/ERROR] [net.minecraft.server.MinecraftServer/]: Failed to reload data packs
java.util.concurrent.ExecutionException: java.lang.RuntimeException: Attempted to load class net/minecraftforge/resource/ReloadRequirements for invalid dist DEDICATED_SERVER
	at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357) ~[?:1.8.0_222]
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1895) ~[?:1.8.0_222]
	at net.minecraft.server.MinecraftServer.func_195568_a(MinecraftServer.java:1467) [?:?]
	at net.minecraft.server.MinecraftServer.func_195560_a(MinecraftServer.java:450) [?:?]
	at net.minecraft.server.MinecraftServer.func_71247_a(MinecraftServer.java:363) [?:?]
	at net.minecraft.server.dedicated.DedicatedServer.func_71197_b(DedicatedServer.java:211) [?:?]
	at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:613) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_222]
Caused by: java.lang.RuntimeException: Attempted to load class net/minecraftforge/resource/ReloadRequirements for invalid dist DEDICATED_SERVER
	at net.minecraftforge.fml.loading.RuntimeDistCleaner.processClass(RuntimeDistCleaner.java:71) ~[forge-1.15.2-31.1.18.jar:31.1]
	at cpw.mods.modlauncher.serviceapi.ILaunchPluginService.processClass(ILaunchPluginService.java:103) ~[modlauncher-5.0.0-milestone.4.jar:5.0.0-milestone.4+67+b1a340b]
	at cpw.mods.modlauncher.LaunchPluginHandler.offerClassNodeToPlugins(LaunchPluginHandler.java:85) ~[modlauncher-5.0.0-milestone.4.jar:?]
	at cpw.mods.modlauncher.ClassTransformer.transform(ClassTransformer.java:115) ~[modlauncher-5.0.0-milestone.4.jar:?]
	at cpw.mods.modlauncher.TransformingClassLoader$DelegatedClassLoader.findClass(TransformingClassLoader.java:239) ~[modlauncher-5.0.0-milestone.4.jar:?]
	at cpw.mods.modlauncher.TransformingClassLoader.loadClass(TransformingClassLoader.java:126) ~[modlauncher-5.0.0-milestone.4.jar:?]
	at cpw.mods.modlauncher.TransformingClassLoader.loadClass(TransformingClassLoader.java:96) ~[modlauncher-5.0.0-milestone.4.jar:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357) ~[?:1.8.0_222]
	at net.minecraftforge.resource.SelectiveReloadStateHandler.get(SelectiveReloadStateHandler.java:64) ~[?:?]
	at net.minecraftforge.resource.ISelectiveResourceReloadListener.func_195410_a(ISelectiveResourceReloadListener.java:33) ~[?:?]
	at net.minecraft.resources.IResourceManagerReloadListener.lambda$reload$0(IResourceManagerReloadListener.java:16) ~[?:?]
	at java.util.concurrent.CompletableFuture.uniRun(CompletableFuture.java:705) ~[?:1.8.0_222]
	at java.util.concurrent.CompletableFuture$UniRun.tryFire(CompletableFuture.java:687) ~[?:1.8.0_222]
	at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:442) ~[?:1.8.0_222]
	at net.minecraft.resources.AsyncReloader.func_219557_a(SourceFile:71) ~[?:?]
	at net.minecraft.util.concurrent.TickDelayedTask.run(SourceFile:18) ~[?:?]
	at net.minecraft.util.concurrent.ThreadTaskExecutor.func_213166_h(SourceFile:144) ~[?:?]
	at net.minecraft.util.concurrent.RecursiveEventLoop.func_213166_h(SourceFile:23) ~[?:?]
	at net.minecraft.server.MinecraftServer.func_213166_h(MinecraftServer.java:731) ~[?:?]
	at net.minecraft.server.MinecraftServer.func_213166_h(MinecraftServer.java:141) ~[?:?]
	at net.minecraft.util.concurrent.ThreadTaskExecutor.func_213168_p(SourceFile:118) ~[?:?]
	at net.minecraft.server.MinecraftServer.func_213205_aW(MinecraftServer.java:714) ~[?:?]
	at net.minecraft.server.MinecraftServer.func_213168_p(MinecraftServer.java:708) ~[?:?]
	at net.minecraft.util.concurrent.ThreadTaskExecutor.func_213161_c(SourceFile:127) ~[?:?]
	at net.minecraft.server.MinecraftServer.func_195568_a(MinecraftServer.java:1464) ~[?:?]
```

Maybe we could annotate ISelectiveResourceReloadListener as client side only as well. This would make it even more clear that it is not server side safe and would cause a more helpful crash.
It does not seem to be referenced in any common/server classes

**Comment by maxanier** [comment]
I would even suggest making the server crash when the data packs cannot be loaded.
Most of MC will work regardless, but some things won't. But it won't be obvious to the end user why things do not work as the small message in the log
`[06Mar2020 06:36:06.861] [Server thread/ERROR] [net.minecraft.server.MinecraftServer/]: Failed to reload data packs`
isn't that clear.

**Comment by sargunv** [comment]
I just ran into this while porting my mod Level Up HP. Forge deprecated MC's `IResourceManagerReloadListener` in favor of this `ISelectiveResourceReloadListener`, but the latter calls `SelectiveReloadStateHandler.get()` which references `ReloadRequirements`, which is client only. Therefore, any data loading (as opposed to asset loading) can't inherit from `ISelectiveResourceReloadListener` as Forge's deprecation comment suggests.

**Comment by diesieben07** [comment]
`IResourceManagerReloadListener` is only used for client-stuff. The server-side data reload listeners (like for example `LootTableManager`) implement the parent `IFutureReloadListener`, which is not depreated.

Tl;dr: Client side use `ISelectiveResourceReloadListener`, server side use `IFutureReloadListener` or one of the classes that implement it.

**Comment by maxanier** [comment]
My main point was that this apparently not clear to mod developers. There isn't any annotation or comment mentioning this.
I spend quite some time debugging the issue linked above, because a) some other mod author did not know about this b) there isn't a way to trace it back to the offending mod (except testing every single mod in a pack of course).
So e.g. your comment could be added somewhere or a crash could be induced somehow which could provide better insight

**Comment by diesieben07** [comment]
Sure, better documentation is always good. Contributions are welcome.

**Comment by maxanier** [comment]
:)
If I want to submit a PR for an issue like this, I have to prepare my workspace, apply the patch, test, create PR with explanation, rebase (including manual patch regeneration) five times before it is merged by Lex (if it ever happens).
This is just to much overhead for such a small change. A member with write permissions could probably implement this in 3 minutes. 
Anyway, I understand that the Forge team does not have time for this.

**Comment by diesieben07** [comment]
That is just BS.
1. This is a forge class, no patch generation needed.
2. A Javadoc PR is much less likely to have merge conflicts.

However that being said, small documentation fixes are usually better suited to be consolidated into one big PR.

**Comment by maxanier** [comment]
yes, you are correct, I exaggerated and in this specific case it would be much easier, sorry.
This is just my general feeling regarding PRs for issues like this or e.g. #6380.

Maybe there could be some label "comment suggestion" and/or "small fix".
Then things like this could be collected and fixed in a bigger PR (or potentially directly by someone with write permissions).
 (Maybe github projects could help as well)

But yes, there are probably more important or fun things the team wants to spend their free time on ;)

**Comment by diesieben07** [comment]
This really isn't the place for this discussion. However regarding the issue you linked: It's corresponding PR got a review within a few days and was merged shortly after. It just took a while for someone to make the PR...
